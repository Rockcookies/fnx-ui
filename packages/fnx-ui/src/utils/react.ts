import {
	Children,
	forwardRef,
	ForwardRefExoticComponent,
	ForwardRefRenderFunction,
	FunctionComponent,
	PropsWithoutRef,
	ReactElement,
	RefAttributes,
} from 'react';
import { isFragment } from 'react-is';

export const toElementArray = (
	children: React.ReactNode,
	option?: { keepEmpty?: boolean },
): ReactElement[] => {
	const keepEmpty = option && option.keepEmpty;
	let ret: React.ReactElement[] = [];

	Children.forEach(children, (child: any) => {
		if ((child === undefined || child === null) && !keepEmpty) {
			return;
		}

		if (Array.isArray(child)) {
			ret = ret.concat(toElementArray(child));
		} else if (isFragment(child) && child.props) {
			ret = ret.concat(toElementArray(child.props.children, option));
		} else {
			ret.push(child);
		}
	});

	return ret;
};

export const createForwardRef = <R, P = any>(
	displayName: string,
	render: ForwardRefRenderFunction<R, P>,
): ForwardRefExoticComponent<PropsWithoutRef<P> & RefAttributes<R>> => {
	const fc = forwardRef(render);
	fc.displayName = displayName;
	return fc;
};

export const createDefaultsForwardRef = <R, P, D>(
	displayName: string,
	defaultProps: D,
	render: ForwardRefRenderFunction<R, P & D>,
): ForwardRefExoticComponent<
	PropsWithoutRef<P & Partial<D>> & RefAttributes<R>
> => {
	const fc = forwardRef(render);
	fc.displayName = displayName;
	fc.defaultProps = defaultProps;
	return fc as any;
};

export const createFC = <P = any>(
	displayName: string,
	fc: FunctionComponent<P>,
): FunctionComponent<P> => {
	fc.displayName = displayName;
	return fc;
};

export const createDefaultsFC = <P, D>(
	displayName: string,
	defaultProps: D,
	fc: FunctionComponent<P & D>,
): FunctionComponent<P & Partial<D>> => {
	fc.displayName = displayName;
	fc.defaultProps = defaultProps;
	return fc as any;
};
