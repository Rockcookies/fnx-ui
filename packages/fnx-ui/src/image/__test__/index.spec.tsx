import { fireEvent, render } from '@testing-library/react';
import React from 'react';
import Image from '../index';

const EXP_IMG = 'https://dummyimage.com/200x200/000/fff.jpg';

describe('<Image/>', () => {
	it('should show correctly when set width,height,src', () => {
		const { container } = render(
			<Image
				data-testid="image"
				width="100"
				height="100"
				src={EXP_IMG}
			/>,
		);

		expect(container).toMatchSnapshot();
	});

	it('should show fit and round style correctly', () => {
		const { container } = render(
			<Image
				data-testid="image"
				round
				fit="cover"
				width="100"
				height="100"
				src={EXP_IMG}
			/>,
		);

		expect(container).toMatchSnapshot();
	});

	it('should show border-radius when radius is set', () => {
		const { container } = render(
			<Image
				data-testid="image"
				radius="10"
				width="100%"
				height="27vw"
				src={EXP_IMG}
			/>,
		);

		expect(container).toMatchSnapshot();
	});
	it('should show alt tip when alt is set', () => {
		const { container } = render(
			<Image
				data-testid="image"
				alt="there is img alt"
				width="100%"
				height="27vw"
				src={EXP_IMG}
			/>,
		);

		expect(container).toMatchSnapshot();
	});

	it('should show define load render', () => {
		const { container } = render(
			<Image
				data-testid="image"
				width="100%"
				height="27vw"
				slots={{
					loading: 'Loading..',
				}}
			/>,
		);
		expect(container).toMatchSnapshot();
	});

	it('should trigger onLoad event', () => {
		const onLoad = jest.fn();

		const { getByTestId } = render(
			<Image
				data-testid="image"
				width="100%"
				height="27vw"
				src={EXP_IMG}
				onLoad={onLoad}
			/>,
		);

		//触发load
		fireEvent(getByTestId('image').children[0], new Event('load'));

		expect(onLoad).toHaveBeenCalled();
	});

	it('should show define error span and trigger onError', () => {
		const onError = jest.fn();

		const { getByTestId, container } = render(
			<Image
				data-testid="image"
				width="100%"
				height="27vw"
				src="x"
				onError={onError}
				slots={{
					error: 'Error',
				}}
			/>,
		);

		//触发error
		fireEvent(getByTestId('image').children[0], new ErrorEvent('error'));
		expect(onError).toHaveBeenCalled();
		expect(container).toMatchSnapshot();
	});

	it('should render loading placeholder when using lazy prop', () => {
		const { container } = render(<Image src={EXP_IMG} lazy />);

		expect(container).toMatchSnapshot();
	});
});
