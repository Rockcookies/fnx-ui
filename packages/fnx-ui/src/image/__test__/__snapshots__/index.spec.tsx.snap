// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Image/> should render loading placeholder when using lazy prop 1`] = `
<div>
  <span
    class="fnx-image"
  >
    <img
      class="fnx-image__img"
      src="https://dummyimage.com/200x200/000/fff.jpg"
      style="object-fit: fill;"
    />
    <span
      class="fnx-image__loading"
    >
      <span
        class="fnx-icon"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="0 0 1024 1024"
          width="1em"
        >
          <path
            d="M312.916992 284.422144c-47.104 0-85.332992 38.228992-85.332992 85.332992s38.228992 85.332992 85.332992 85.332992 85.334016-38.228992 85.334016-85.332992-38.230016-85.332992-85.334016-85.332992m356.665344 51.10784c-22.868992-26.73664-64.340992-26.452992-86.868992 0.626688L341.332992 625.77664l-73.499648-73.556992c-22.244352-22.243328-58.25536-22.243328-80.498688 0l-73.556992 73.556992V227.555328H910.22336v388.721664l-240.64-280.745984z m240.64-164.864H113.77664c-31.403008 0-56.88832 25.430016-56.88832 56.889344v568.88832c0 31.403008 25.485312 56.889344 56.88832 56.889344H910.22336c31.459328 0 56.88832-25.486336 56.88832-56.88832V227.555328c0-31.459328-25.428992-56.88832-56.88832-56.88832z"
          />
        </svg>
      </span>
    </span>
  </span>
</div>
`;

exports[`<Image/> should show alt tip when alt is set 1`] = `
<div>
  <span
    class="fnx-image"
    data-testid="image"
    style="width: 100%; height: 27vw;"
  >
    <img
      alt="there is img alt"
      class="fnx-image__img"
      src="https://dummyimage.com/200x200/000/fff.jpg"
      style="object-fit: fill;"
    />
    <span
      class="fnx-image__loading"
    >
      <span
        class="fnx-icon"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="0 0 1024 1024"
          width="1em"
        >
          <path
            d="M312.916992 284.422144c-47.104 0-85.332992 38.228992-85.332992 85.332992s38.228992 85.332992 85.332992 85.332992 85.334016-38.228992 85.334016-85.332992-38.230016-85.332992-85.334016-85.332992m356.665344 51.10784c-22.868992-26.73664-64.340992-26.452992-86.868992 0.626688L341.332992 625.77664l-73.499648-73.556992c-22.244352-22.243328-58.25536-22.243328-80.498688 0l-73.556992 73.556992V227.555328H910.22336v388.721664l-240.64-280.745984z m240.64-164.864H113.77664c-31.403008 0-56.88832 25.430016-56.88832 56.889344v568.88832c0 31.403008 25.485312 56.889344 56.88832 56.889344H910.22336c31.459328 0 56.88832-25.486336 56.88832-56.88832V227.555328c0-31.459328-25.428992-56.88832-56.88832-56.88832z"
          />
        </svg>
      </span>
    </span>
  </span>
</div>
`;

exports[`<Image/> should show border-radius when radius is set 1`] = `
<div>
  <span
    class="fnx-image"
    data-testid="image"
    style="width: 100%; height: 27vw; overflow: hidden; border-radius: 10px;"
  >
    <img
      class="fnx-image__img"
      src="https://dummyimage.com/200x200/000/fff.jpg"
      style="object-fit: fill;"
    />
    <span
      class="fnx-image__loading"
    >
      <span
        class="fnx-icon"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="0 0 1024 1024"
          width="1em"
        >
          <path
            d="M312.916992 284.422144c-47.104 0-85.332992 38.228992-85.332992 85.332992s38.228992 85.332992 85.332992 85.332992 85.334016-38.228992 85.334016-85.332992-38.230016-85.332992-85.334016-85.332992m356.665344 51.10784c-22.868992-26.73664-64.340992-26.452992-86.868992 0.626688L341.332992 625.77664l-73.499648-73.556992c-22.244352-22.243328-58.25536-22.243328-80.498688 0l-73.556992 73.556992V227.555328H910.22336v388.721664l-240.64-280.745984z m240.64-164.864H113.77664c-31.403008 0-56.88832 25.430016-56.88832 56.889344v568.88832c0 31.403008 25.485312 56.889344 56.88832 56.889344H910.22336c31.459328 0 56.88832-25.486336 56.88832-56.88832V227.555328c0-31.459328-25.428992-56.88832-56.88832-56.88832z"
          />
        </svg>
      </span>
    </span>
  </span>
</div>
`;

exports[`<Image/> should show correctly when set width,height,src 1`] = `
<div>
  <span
    class="fnx-image"
    data-testid="image"
    style="width: 100px; height: 100px;"
  >
    <img
      class="fnx-image__img"
      src="https://dummyimage.com/200x200/000/fff.jpg"
      style="object-fit: fill;"
    />
    <span
      class="fnx-image__loading"
    >
      <span
        class="fnx-icon"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="0 0 1024 1024"
          width="1em"
        >
          <path
            d="M312.916992 284.422144c-47.104 0-85.332992 38.228992-85.332992 85.332992s38.228992 85.332992 85.332992 85.332992 85.334016-38.228992 85.334016-85.332992-38.230016-85.332992-85.334016-85.332992m356.665344 51.10784c-22.868992-26.73664-64.340992-26.452992-86.868992 0.626688L341.332992 625.77664l-73.499648-73.556992c-22.244352-22.243328-58.25536-22.243328-80.498688 0l-73.556992 73.556992V227.555328H910.22336v388.721664l-240.64-280.745984z m240.64-164.864H113.77664c-31.403008 0-56.88832 25.430016-56.88832 56.889344v568.88832c0 31.403008 25.485312 56.889344 56.88832 56.889344H910.22336c31.459328 0 56.88832-25.486336 56.88832-56.88832V227.555328c0-31.459328-25.428992-56.88832-56.88832-56.88832z"
          />
        </svg>
      </span>
    </span>
  </span>
</div>
`;

exports[`<Image/> should show define error span and trigger onError 1`] = `
<div>
  <span
    class="fnx-image"
    data-testid="image"
    style="width: 100%; height: 27vw;"
  >
    <span
      class="fnx-image__error"
    >
      Error
    </span>
  </span>
</div>
`;

exports[`<Image/> should show define load render 1`] = `
<div>
  <span
    class="fnx-image"
    data-testid="image"
    style="width: 100%; height: 27vw;"
  >
    <span
      class="fnx-image__loading"
    >
      Loading..
    </span>
  </span>
</div>
`;

exports[`<Image/> should show fit and round style correctly 1`] = `
<div>
  <span
    class="fnx-image fnx-image--round"
    data-testid="image"
    style="width: 100px; height: 100px;"
  >
    <img
      class="fnx-image__img"
      src="https://dummyimage.com/200x200/000/fff.jpg"
      style="object-fit: cover;"
    />
    <span
      class="fnx-image__loading"
    >
      <span
        class="fnx-icon"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="0 0 1024 1024"
          width="1em"
        >
          <path
            d="M312.916992 284.422144c-47.104 0-85.332992 38.228992-85.332992 85.332992s38.228992 85.332992 85.332992 85.332992 85.334016-38.228992 85.334016-85.332992-38.230016-85.332992-85.334016-85.332992m356.665344 51.10784c-22.868992-26.73664-64.340992-26.452992-86.868992 0.626688L341.332992 625.77664l-73.499648-73.556992c-22.244352-22.243328-58.25536-22.243328-80.498688 0l-73.556992 73.556992V227.555328H910.22336v388.721664l-240.64-280.745984z m240.64-164.864H113.77664c-31.403008 0-56.88832 25.430016-56.88832 56.889344v568.88832c0 31.403008 25.485312 56.889344 56.88832 56.889344H910.22336c31.459328 0 56.88832-25.486336 56.88832-56.88832V227.555328c0-31.459328-25.428992-56.88832-56.88832-56.88832z"
          />
        </svg>
      </span>
    </span>
  </span>
</div>
`;
